### PR TITLE
system.tables improvements (total_rows/total_bytes/storage_policy)

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `total_rows`/`total_bytes` into the `system.tables` table.

Detailed description / Documentation draft:

#### This patch-set contains the following changes:
- set `storage_policy` for `Distributed` table engine, and hence showing it in `system.tables` (follow up for #8756)
- `total_rows` for `*MergeTree/Buffer/Memory` in `system.tables`
- `total_bytes` for `*MergeTree/Buffer/Memory` in `system.tables`

This differs from the `system.parts` since it adds rows/bytes for `Buffer`/`Memory` too, which can be used to track memory usage (since the memory that is used for `Buffer`/`Memory` rows does not displays anywhere AFAICS)

*Please note, that `total_rows` includes rows of the underlying storage for `Buffer` engine (since it uses `IStorage::totalRows()` which used for `SELECT count() FROM table` trivial queries), while `total_bytes` does not (since otherwise the purpose of this column will be broken).*

#### Also some issues which was found during implementing this patch set has been fixed to:

- `optimize_trivial_count_query` with `GROUP BY`

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Thank you! total rows in system.tables is very useful!

File: dbms/src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -2995,9 +2995,9 @@ Pipes StorageReplicatedMergeTree::read(
 }
 
 
-std::optional<UInt64> StorageReplicatedMergeTree::totalRows() const
+template <class Func>
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageReplicatedMergeTree.cpp]: Committed (double m, double t)

**Review by alexey-milovidov — APPROVED** [review]

**Comment by azat** [comment]
> 00341_squashing_insert_select2

This is due to incorrect detection of `optimize_trivial_count_query`, fixed too (can be moved into separate PR, since it is completely unrelated to this one, but it's up to you)

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
It's fine :)

**Comment by azat** [comment]
Ok, PR description updated (to reflect this info at least).

**Comment by azat** [comment]
>Integration tests (asan) — failed:1, passed:0, error:0

`==================== 2 passed, 2 warnings in 76.88 seconds =====================`

Hm, interesting.

**Comment by alexey-milovidov** [comment]
Minor:
```
2020-03-29 17:18:04 /build/obj-x86_64-linux-gnu/../dbms/src/Storages/StorageBuffer.cpp:730:27: error: invalid case style for local variable 'underlyingRows' [readability-identifier-naming,-warnings-as-errors]
2020-03-29 17:18:04     std::optional<UInt64> underlyingRows;
2020-03-29 17:18:04                           ^~~~~~~~~~~~~~
2020-03-29 17:18:04                           underlying_rows
```

**Comment by azat** [comment]
>readability-identifier-naming

Interesting, why it does not report the same for totalBytes and totalRows

**Comment by azat** [comment]
So the following tests fails now:
- Integration tests -- fails in upstream/master too
- `Functional stateless tests (ubsan) — fail: 4` -- fails in upstream/master too (all 4 matched)
- `01098_temporary_and_external_tables` -- looks "flacky", since it does not fails in previous version of this patch (and it passed locally)

**Comment by alexey-milovidov** [comment]
A bug has been found: https://github.com/ClickHouse/ClickHouse/issues/40637

**Pull request merged**